File: eclipse/src/utilities/adt/moe/BinNode.java
package utilities.adt.moe;

/** This class represents a binary tree node.
 * 
 * @author MOE
 *
 * @param <T> The type of object stored within the node.
 */
public class BinNode<T> {
	
	/** The value of the current binary tree node. */
    private T value;
    
    /** The left child (node) of the current binary tree node. */
    private BinNode<T> left;
    
    /** The right child (node) of the current binary tree node. */ 
    private BinNode<T> right;

    /** C'tor which gets the value to store in the node.
     * @param value The value of the newly created binary tree node.
     */
    public BinNode(T value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
    
    /** C'tor which gets values for this node and the 2 nodes which are its children.
     * @param left The left child node.
     * @param value The value of this new node.
     * @param right The right child node.
     */
    public BinNode(BinNode<T> left, T value, BinNode<T> right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    /** Return true if this node has a left child, otherwise - false. */
    public boolean hasLeft() {
        return left != null ;
    }

    /** Return true if this node has a right child, otherwise - false. */
    public boolean hasRight() {
        return right != null ;
    }

    /** Returns the value (type {@link T}) of this binary tree node. */
    public T getValue() {
        return value;
    }
    
    /** Returns the left child of this binary tree node. */
    public BinNode<T> getLeft() {
        return left;
    }
    
    /** Returns the right child of this binary tree node. */
    public BinNode<T> getRight() {
        return right;
    }
    
    /** Updates the value of this binary tree node. */
    public void setValue(T value) {
        this.value = value;
    }
    
    /** Updates the left child of this binary tree node. */
    public void setLeft(BinNode<T> left) {
        this.left = left;
    }
    
    /** Updates the right child of this binary tree node. */
    public void setRight(BinNode<T> right) {
        this.right = right;
    }

    /** Returns the {@link String} representation of the binary sub-tree,
     * starting with this node. */
    public String toString() {
        return " ( " + left + " " + value + " " + right + " ) ";
    }
}

File: VS/utils/BinNode/BinNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MOE.BinNode {

	/// <summary>
	/// This class represents a node which stores a (generic) value and 2 children<br/>
	/// which are similar nodes.<br/>
	/// <br/>
	/// We can use this class to represent a bi-directional (doubly) linked list or a binary tree.
	/// </summary>
	public class BinNode<T> {
		
		/// <summary>
		/// Reference to the left child (or `prev` in a bi-directional linked list).
		/// </summary>
		private BinNode<T> left;
	
		/// <summary>
		/// The value stored in this node.
		/// </summary>
		private T value;
		
		/// <summary>
		/// Reference to the left child (or `prev` in a bi-directional linked list).
		/// </summary>
		private BinNode<T> right;
		
		/// <summary>
		/// A c'tor which creates a tree node (instance of this class),
		/// its `left` and `right` properties refer to null.
		/// </summary>
		/// <param name="x">The value of the created instance.</param>
		public BinNode(T x) {
			this.left = null;
			this.value = x;
			this.right = null;
		}
		
		/// <summary>
		/// A c'tor which creates a tree node (instance of this class),
		/// its properties would get the values of the arguments.
		/// </summary>
		/// <param name="left"></param>
		/// <param name="x"></param>
		/// <param name="right"></param>
		public BinNode(BinNode<T> left, T x, BinNode<T> right) {
			this.left = left;
			this.value = x;
			this.right = right;
		}
		
		/// <summary>
		/// Returns the value stored in this tree node.
		/// </summary>
		/// <returns></returns>
		public T GetValue() {
			return this.value;
		}
		
		/// <summary>
		/// Updates the value stored in the current tree node.
		/// </summary>
		/// <param name="x"></param>
		public void SetValue(T x) {
			this.value = x;
		}
		
		/// <summary>
		/// Returns this node's left child.
		/// </summary>
		/// <returns></returns>
		public BinNode<T> GetLeft() {
			return this.left;
		}
		
		/// <summary>
		/// Returns this node's right child.
		/// </summary>
		/// <returns></returns>
		public BinNode<T> GetRight() {
			return this.right;
		}
		
		/// <summary>
		/// Updates the left child of the current tree node.
		/// </summary>
		/// <param name="left"></param>
		public void SetLeft(BinNode<T> left) {
			this.left = left;
		}
		
		/// <summary>
		/// Updates the right child of the current tree node.
		/// </summary>
		/// <param name="right"></param>
		public void SetRight(BinNode<T> right) {
			this.right = right;
		}
		
		/// <summary>
		/// Returns if the current node has a left child.
		/// </summary>
		/// <returns></returns>
		public bool HasLeft() {
			return this.left != null;
		}
		
		/// <summary>
		/// Returns if the current node has a right child.
		/// </summary>
		/// <returns></returns>
		public bool HasRight() {
			return this.right != null;
		}
		
		/// <summary>
		/// Returns the string representation of the property `value` in the current node.
		/// </summary>
		/// <returns></returns>
		public override string ToString() {
			return this.value.ToString();
		}
	}
}